### PR TITLE
Enhance GET _node/stats/pipelines API for Metricbeat monitoring

### PR DESCRIPTION
**Pull request by cachedout**
Closes https://github.com/elastic/logstash/issues/10120

This will fail tests until https://github.com/elastic/logstash/pull/10561 is merged in and this PR is rebased. I'm just parking it here until then. ;)

### CODE REVIEW

**Comment by ycombinator** [comment]
@cachedout Looks like this PR now needs a merge/rebase from `master`.

**Comment by cachedout** [comment]
@jsvd and @yaauie I believe the only remaining test failure is unrelated to this PR so I consider it ready for review. @ycombinator is currently doing some functional testing on it as well.

**Comment by ycombinator** [comment]
Hi @cachedout, I noticed a few things after running this PR:

1. Per https://github.com/elastic/logstash/issues/10120#issuecomment-439458354, the `vertices` field should only be returned if `?vertices=true` is set. At the moment the `vertices` field is always returned in the API response.

2. If you look at the sample JSON document in https://github.com/elastic/logstash/issues/10120#issue-377787260, particularly the `queue` field, you'll notice it looks like this:

   ```
   queue":{
     "type":"memory",
     "queue_size_in_bytes":0,
     "max_queue_size_in_bytes":0,
     "events_count":0
   }
   ```

   However, the `queue` field in the API response only contains the `type` field under it.

3. This is a new requirement so feel free to defer it to a follow up PR. If the `elasticsearch` output plugin is being used in a pipeline, could we add `cluster_uuids` to that pipeline's object in the API response, similar to what we did in `GET /_node/pipelines` API response?

**Comment by ycombinator** [comment]
@cachedout I pulled down this PR and re-tested it. Looks like points 1 and 2 from my previous comment have been addressed but point 3 hasn't yet.

**Comment by cachedout** [comment]
@ycombinator Sorry for the confusion. I was going to address the third point in a follow-up PR.

**Comment by cachedout** [comment]
This is ready for a review from @jsvd or @yaauie or somebody else on @elastic/logstash

**Comment by danhermann** [comment]
We are going to need to commit the changes from #10561 into master since they were merged only into the 7.x branch. Is there a plan for addressing the tech debt introduced in #10561?

**Comment by cachedout** [comment]
@danhermann By "tech debt" are you referring to [this comment](https://github.com/elastic/logstash/pull/10561#issuecomment-489663471)?

**Comment by danhermann** [comment]
> @danhermann By "tech debt" are you referring to [this comment](https://github.com/elastic/logstash/pull/10561#issuecomment-489663471)?

Both that one and the earlier comment about tighter coupling than we would like between Logstash core and plugins.

**Comment by cachedout** [comment]
Hi @danhermann !

Forgive me, but I'm a bit confused by this:

> We are going to need to commit the changes from #10561 into master since they were merged only into the 7.x branch.

The [changes made by the commits in that PR ](https://github.com/elastic/logstash/commit/cc3c5ec00f76e31e2f494774a30a5bd6c0afee68)do seem to be present in the master branch. For example, [see the git blame here](https://github.com/elastic/logstash/blame/master/logstash-core/lib/logstash/api/commands/node.rb) which seems to show these commits present in the master branch. Just in case, I tried creating a PR for but while the commits themselves apply cleanly, [such a PR seems to result in no proposed changes to the files in the Logstash master branch](https://github.com/elastic/logstash/compare/master...cachedout:fwd_merge_10561?expand=1).

Regarding the comments about tight coupling, my understanding is that those were resolved via the introduction of [the PluginMetadata system which was merged here](https://github.com/elastic/logstash/pull/10691) and subsequent use of the new system in my changes. Did I misunderstand the nature of the tight-coupling problem? If you could provide more specifics about the nature of this problem which remain unresolved, I'd be happy of course to try and resolve them. :)

**Comment by danhermann** [comment]
> The [changes made by the commits in that PR ](https://github.com/elastic/logstash/commit/cc3c5ec00f76e31e2f494774a30a5bd6c0afee68)do seem to be present in the master branch. 

My apologies, @cachedout. I saw those changes first as a series of distinct commits in the 7.x branch and missed that they had been committed into master as a single, squashed commit (as we prefer).

> Regarding the comments about tight coupling, my understanding is that those were resolved via the introduction of [the PluginMetadata system which was merged here](https://github.com/elastic/logstash/pull/10691) and subsequent use of the new system in my changes. Did I misunderstand the nature of the tight-coupling problem? If you could provide more specifics about the nature of this problem which remain unresolved, I'd be happy of course to try and resolve them. :)

Let me take another look at those and get back to you in a day or two if that's ok.

**Comment by cachedout** [comment]
>  Please let me know if you plan to address it in this PR or a follow up PR.

@ycombinator The functionality to include cluster uuids in ES outputters is now a part of this PR. Please have a look and see if it suits your needs.

**Comment by ycombinator** [comment]
@cachedout It looks like the `cluster_uuids` field shows up under `pipelines.{id}.plugins.outputs[n].cluster_uuids` when the `GET _node/stats/pipelines` API is called, where `n` is the index of to an elasticsearch output plugin object. Contrast this with the response from the `GET _node/pipelines` API, where `cluster_uuids` shows up under `pipelines.{id}.cluster_uuids`, i.e. not scoped specifically under an elasticsearch plugin or vertex object.

I think the two APIs responses should be consistent in where `cluster_uuids` shows up. Conceptually, I prefer the direction you've taken with the `GET _node/stats/pipelines` API in this PR, w.r.t. the location of the `cluster_uuids` field. Specifically, I like that it's within the scope of the Elasticsearch output object in the API response. 

So I'd suggest following the same pattern in the `GET _node/pipelines` API response as well. Of course, this API doesn't have a `pipelines.{id}.plugins` object. Both API responses, however, have a `pipelines.{id}.**.vertices` object. Would it be possible, in both API responses, to inject the `cluster_uuid` (note singluar) property under the vertex object corresponding to the Elasticsearch output?

**Comment by cachedout** [comment]
@ycombinator OK, I went ahead and made that change. I left vertices where a lookup for cluster uuids fails as null so you don't have to check for the existence of a key. Let me know if you'd like that to change. Otherwise, if this looks OK then let me know and I'll make the changes to the other endpoint to match.

**Comment by ycombinator** [comment]
@cachedout I like it! The `null` value when the cluster UUID can't be determined is fine.

One minor change: now that we're scoping the cluster UUIDs to individual Elastiscearch output(s), should we rename the field in the API response to `cluster_uuid` (singular)?

**Comment by ycombinator** [comment]
Whoa, hang on. I just noticed that all vertices, regardless of whether they have anything to do with Elasticsearch or not, are now getting a `cluster_uuids` property. This doesn't seem right to me. Is there no way to give this property to only vertices that represent Elasticsearch output plugins?

**Comment by cachedout** [comment]
@ycombinator Yes, this is what I meant when I said that vertices where lookups failed have a null entry. I can remove them. Sorry if that wasn't clear. I can remove them. :)

**Comment by cachedout** [comment]
@ycombinator OK, ready for you to review again. :)

**Comment by ycombinator** [comment]
I tested this PR with the following pipelines:

* `input { stdin {} }`
   ✔️ two vertices, neither show `cluster_uuids` field

* `input { stdin {} } output { stdout {} }`
   ✔️ two vertices, neither show `cluster_uuids` field

* `input { stdin {} } output { elasticsearch {} }`
   ✔️ two vertices, only the ES output vertex shows `cluster_uuids` field with the right value

* `input { stdin {} } output { elasticsearch {} elasticsearch {} }`
   ✔️ three vertices, each of the two ES output vertices show `cluster_uuids` field with the right value

* `input { stdin {} } output { elasticsearch {} elasticsearch { hosts => [ "localhost:9210" ] } }`
   ✔️ three vertices, each of the two ES output vertices show `cluster_uuids` field and the field has the right value

* `input { stdin {} } filter { elasticsearch {} } output { elasticsearch {} }`
   ✔️ three vertices, only the ES output vertex shows `cluster_uuids` field and the field has the right value

Functionally, things look very good to me now. Two questions:

1. What do you think of changing the field name from `cluster_uuids` to `cluster_uuid`, now that the field is scoped to an individual Elasticsearch plugin vertex?

2. Do you think it's worth writing tests using each of the pipelines I used above in my manual testing? I don't really know the implementation details here so it's possible there are already tests elsewhere that might be sufficient. I also don't know how easy it would be to write such tests.

**Comment by cachedout** [comment]
> What do you think of changing the field name from cluster_uuids to cluster_uuid, now that the field is scoped to an individual Elasticsearch plugin vertex?

Derp. You mentioned that earlier and I forgot when I was changing the other thing. Will fix.

> Do you think it's worth writing tests using each of the pipelines I used above in my manual testing? I don't really know the implementation details here so it's possible there are already tests elsewhere that might be sufficient. I also don't know how easy it would be to write such tests.

I have come across tests for these but I don't have a good feel for how easy they are to modify. I do think it's important to have test coverage here though so I'll investigate. It's up to the logstash team as far as I'm concerned to decide whether to accept this PR without tests or if they want them first. Either is fine with me.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 1]
```diff
@@ -110,8 +130,14 @@ def plugin_stats(stats, plugin_type)
             end
           end
 
-          def report(stats)
-            {
+          def report(stats, extended_stats, opts)
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: `PluginStats#report` is a public method, so I'd rather avoid breaking its signature; try:

```suggestion
          def report(stats, extended_stats=nil, opts={})
```

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 2]
```diff
@@ -92,9 +108,13 @@ def hot_threads(options={})
         end
 
         private
-        def plugins_stats_report(pipeline_id)
+        def plugins_stats_report(pipeline_id, opts)
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: This method is private, so we _should_ be able to find and modify all call-sites, but a non-breaking equivalent signature would be:

```suggestion
        def plugins_stats_report(pipeline_id, opts={})
```

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 3]
```diff
@@ -45,14 +61,14 @@ def events
           )
         end
 
-        def pipeline(pipeline_id = nil)
+        def pipeline(pipeline_id = nil, opts)
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: Since `pipeline` is a public method, I would prefer we not change its signature in a breaking way since there may be other bits that rely on the ability to send `Stats#pipeline` with no arguments.

```suggestion
        def pipeline(pipeline_id = nil, opts={})
```

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 4]
```diff
@@ -120,9 +146,20 @@ def report(stats)
...
-              :queue => stats[:queue]
+              :queue => extended_stats["queue"]
             }.merge(stats[:dlq] ? {:dead_letter_queue => stats[:dlq]} : {})
+            if !opts[:vertices]
+              ret.delete('vertices')
+              unless extended_stats.nil?
+                extended_stats.delete('vertices')
+              end
+            end
+            if extended_stats.nil?
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: `extended_stats` will never be nil here, since we send `Hash#[]` to it in your change on line 149 which would raise a `NoMethodError`.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 5]
```diff
@@ -120,9 +146,20 @@ def report(stats)
                 :outputs => plugin_stats(stats, :outputs)
               },
               :reloads => stats[:reloads],
-              :queue => stats[:queue]
+              :queue => extended_stats["queue"]
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: is this change intentional? why do we need to reach into `extended_stats["queue"]` for this value, if it has historically been available on `stats[:queue]`?

File: logstash-core/lib/logstash/config/pipelines_info.rb [hunk 6]
```diff
@@ -13,6 +13,14 @@ def self.format_pipelines_info(agent, metric_store, extended_performance_collect
         p_stats = stats[pipeline_id]
         # Don't record stats for system pipelines
         next nil if pipeline.system?
+        i = 0
+        for output in p_stats[:plugins][:outputs]
+          plugin_id = output[i].to_s
+          if LogStash::PluginMetadata.exists?(plugin_id)
+            cluster_uuids = LogStash::PluginMetadata.for_plugin(plugin_id).get(:cluster_uuid)
```
Comment (yaauie) [inline logstash-core/lib/logstash/config/pipelines_info.rb]: this local variable is assigned, but not aver accessed before falling out of scope, which leads me to believe that this entire hunk is not necessary.

File: logstash-core/lib/logstash/config/pipelines_info.rb [hunk 7]
```diff
@@ -93,11 +99,19 @@ def self.format_pipeline_vertex_section_stats(stats, pipeline)
 
           acc
         end
-        
-        acc << {
+        if LogStash::PluginMetadata.exists?(plugin_id.to_s)
+          cluster_uuids = LogStash::PluginMetadata.for_plugin(plugin_id.to_s).get(:cluster_uuid)
```
Comment (yaauie) [inline logstash-core/lib/logstash/config/pipelines_info.rb]: nitpick: since this now holds at most one uuid, the variable name should be singular `cluster_uuid`.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 8]
```diff
@@ -92,9 +108,13 @@ def hot_threads(options={})
...
         private
-        def plugins_stats_report(pipeline_id)
+        def plugins_stats_report(pipeline_id, opts)
           stats = service.get_shallow(:stats, :pipelines, pipeline_id.to_sym)
-          PluginsStats.report(stats)
+          extended_stats = LogStash::Config::PipelinesInfo.format_pipelines_info(
+            service.agent,
+            service.snapshot.metric_store,
+            true)
+          PluginsStats.report(stats, extended_stats.shift, opts)
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: It is unclear to me what we are attempting to accomplish with this `Array#shift` sent to `extended_stats`.

From my reading, `LogStash::Config::PipelinesInfo::format_pipelines_info` is returning an array of data for _all_ of the running pipelines, so returning the first element is not guaranteed to pick off the extended stats that match the `pipeline_id` that has been passed to this method.

File: logstash-core/lib/logstash/config/pipelines_info.rb (hunk 7 above)
Comment (cachedout) [reply logstash-core/lib/logstash/config/pipelines_info.rb]: Good catch. Thanks! Fixed.

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 4 above)
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Thanks! Removed.

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 8 above)
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Gah! Good catch. I had put this in for testing and forgot to remove it. I changed it so that we should now pass in the correct pipeline.

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 5 above)
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: This change originates from [this requirement](https://github.com/elastic/logstash/pull/10576#issuecomment-491331012).

**Comment by cachedout** [comment]
Hi @yaauie. Thanks for the review! I believe I have now addressed your concerns. Could you please have another look?

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 9]
```diff
@@ -110,8 +135,14 @@ def plugin_stats(stats, plugin_type)
             end
           end
 
-          def report(stats)
-            {
+          def report(stats, extended_stats=nil, opts={})
```
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: I applied the suggestion to fall through on L154 and so will resolve this comment if that's all right.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 10]
```diff
@@ -120,9 +151,16 @@ def report(stats)
                 :outputs => plugin_stats(stats, :outputs)
               },
               :reloads => stats[:reloads],
-              :queue => stats[:queue]
+              :queue => extended_stats["queue"]
             }.merge(stats[:dlq] ? {:dead_letter_queue => stats[:dlq]} : {})
+            if !opts[:vertices]
+              ret.delete('vertices')
+              unless extended_stats.nil?
```
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Now that we have changed L154, and given that the interface still defaults `extended_stats` to nil, do you still feel like we should remove this check?

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 11]
```diff
@@ -110,8 +135,14 @@ def plugin_stats(stats, plugin_type)
             end
           end
 
-          def report(stats)
-            {
+          def report(stats, extended_stats=nil, opts={})
+            if !stats[:config].nil?
```
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: This was part of doing some debugging from a previous change and I failed to remove it. I have done so.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 12]
```diff
@@ -92,9 +108,18 @@ def hot_threads(options={})
         end
 
         private
-        def plugins_stats_report(pipeline_id)
+        def plugins_stats_report(pipeline_id, opts={})
           stats = service.get_shallow(:stats, :pipelines, pipeline_id.to_sym)
-          PluginsStats.report(stats)
+          extended_stats = LogStash::Config::PipelinesInfo.format_pipelines_info(
```
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Good catch! I changed this to your suggestion.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 13]
```diff
@@ -92,9 +108,18 @@ def hot_threads(options={})
...
+            service.agent,
+            service.snapshot.metric_store,
+            true)
+          for pipeline in extended_stats
```
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: I believe this should now be resolved by resolving the complexity elsewhere. I am setting this to resolved. Please let me know if you feel like it needs to be addressed.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 14]
```diff
@@ -45,14 +60,22 @@ def events
           )
         end
 
-        def pipeline(pipeline_id = nil)
+        def pipeline(pipeline_id = nil, opts={})
           if pipeline_id.nil?
             pipeline_ids = service.get_shallow(:stats, :pipelines).keys
+            extended_stats = LogStash::Config::PipelinesInfo.format_pipelines_info(
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: since we need the `extended_stats` either way, let's bring this up above the `if` clause so we can avoid repetition.

Also, as I [previously suggested](https://github.com/elastic/logstash/pull/10576#pullrequestreview-248312786), indexing the extended stats into a `Hash{String=>Hash}` will incur a small cost once that will drastically reduce the repetitive cost of lookups.

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (yaauie) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Please see [this comment](https://github.com/elastic/logstash/pull/10576#discussion_r292573616), which has a suggested path forward. We are adding a lot of complexity by merging our carefully-crafted `ret` into `extended_stats` and then attempting to remove the bits we don't want, and it would be simpler to explicitly compose the bits that we do want.

Doing so would also shield us from needing to mutate the objects that were passed to us.

File: logstash-core/lib/logstash/config/pipelines_info.rb [hunk 15]
```diff
@@ -93,11 +91,19 @@ def self.format_pipeline_vertex_section_stats(stats, pipeline)
 
           acc
         end
-        
-        acc << {
+        if LogStash::PluginMetadata.exists?(plugin_id.to_s)
```
Comment (yaauie) [inline logstash-core/lib/logstash/config/pipelines_info.rb]: nitpick: splitting up the building of `cluster_uuid` variable and its use is a bit harder to follow than it needs to be. I think that the following is effectively identical, but is a little easier to follow.

~~~ ruby
        segment = {
          :id => plugin_id,
          :pipeline_ephemeral_id => pipeline.ephemeral_id
        }

        if LogStash::PluginMetadata.exists?(plugin_id.to_s)
          plugin_metadata = LogStash::PluginMetadata.for_plugin(plugin_id.to_s)
          cluster_uuid = plugin_metadata&.get(:cluster_uuid)
          segment[:cluster_uuid] = cluster_uuid unless cluster_uuid.nil?
        end

        acc << segment.merge(segmented)
~~~

Comment (cachedout) [reply logstash-core/lib/logstash/config/pipelines_info.rb]: Thanks! I made this change as recommended.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 16]
```diff
@@ -120,9 +151,16 @@ def report(stats)
...
-              :queue => stats[:queue]
+              :queue => extended_stats["queue"]
             }.merge(stats[:dlq] ? {:dead_letter_queue => stats[:dlq]} : {})
+            if !opts[:vertices]
+              ret.delete('vertices')
+              unless extended_stats.nil?
+                extended_stats.delete('vertices')
+              end
             end
+          extended_stats.merge(ret)
```
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: I agree that this is much better. Thanks! I have implemented this change.

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: I have implemented your improvements to the way we compose that return. Therefore, I'm going to make this as resolved.

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 14 above)
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: I think I've resolved this concern now. I did repeat the logic a bit so there is some room for DRY improvements but I thought this might be more readable. Happy to change it though if you think it would be cleaner another way. :)

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 13 above)
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: I think this should be resolved now.

**Comment by cachedout** [comment]
@yaauie Thanks very much for the additional review! I believe I have addressed the concerns you raised. Could you please take another look?

@ycombinator The `cluster_uuid` field has been added [as you requested here](https://github.com/elastic/logstash/pull/10576#issuecomment-498516575).

**Comment by ycombinator** [comment]
> The `cluster_uuid` field has been added [as you requested here](https://github.com/elastic/logstash/pull/10576#issuecomment-498516575).

Thanks, I see it in the `GET _node` API response now, under elasticsearch output plugin vertices.

However, I see that the older `pipelines.{id}.cluster_uuids` field is still present as well. Did you forget to remove this or did you leave it as-is intentionally?

**Comment by cachedout** [comment]
I wasn't sure if it would still be useful so I left it in. I've just pushed a commit to remove it though.

**Comment by ycombinator** [comment]
> I wasn't sure if it would still be useful so I left it in. 

I can see both sides of it:

- If we leave it in, it's backwards compatible.
- If we take it out, its a cleaner API.

Personally I'm okay with taking it out because I _suspect_ the Metricbeat `logstash` module is the only thing actually using it at the moment.

**Comment by cachedout** [comment]
> Personally I'm okay with taking it out because I suspect the Metricbeat logstash module is the only thing actually using it at the moment.

Yes, that is what I assume as well. Though, I'll do whatever @elastic/logstash thinks is best. :)

**Comment by cachedout** [comment]
@ycombinator Oh, there is one issue that I wanted to bring up. Right now, to display vertices there are two possible flags -- `?vertices=true` and `?graph=true`, depending on which endpoint you call. I think those should probably be consistent. Do you agree and if so, do you have a preference?

**Comment by ycombinator** [comment]
@cachedout IMO it'd be nice if the query string parameter name was consistent. My preference would be for `graph`.

**Comment by ycombinator** [comment]
@cachedout Some change in this PR seems to have broken the functionality of the `graph` query string parameter on the `GET _node/pipelines` API, introduced in https://github.com/elastic/logstash/pull/10561.

On `master` (i.e. without running Logstash with the changes in this PR), if I call the `GET _node/pipelines` API without the `graph` query string parameter, I don't get the `graph` field in the API response, as expected.

However, with this PR, if I call the `GET _node/pipelines` API without the `graph` query string parameter, I get the `graph` field in the API response anyway.

File: logstash-core/lib/logstash/config/lir_serializer.rb [hunk 17]
```diff
@@ -64,7 +64,18 @@ def vertex_type(v)
...
     
+    def get_uuid(plugin_id)
+      if LogStash::PluginMetadata.exists?(plugin_id.to_s)
+        plugin_metadata = LogStash::PluginMetadata.for_plugin(plugin_id.to_s)
+        cluster_uuid = plugin_metadata&.get(:cluster_uuid)
+      end
+    end
+    
     def decorate_vertex(v, v_json)
+      uuid = get_uuid(v.id)
```
Comment (yaauie) [inline logstash-core/lib/logstash/config/lir_serializer.rb]: this appears to be an entirely new change, so I feel like I'm reviewing a moving target here; what brought about the requirement to add this to the LIR serialisation? The LIR is an internal representation of the config's AST as compiled by the grammar, and injecting runtime metadata into it feels like a recipe for surprises later down the road.

**Comment by ycombinator** [comment]
Apologies @yaauie, for making this PR a moving target in terms of functionality. We know the amount of time you can give here is limited so we certainly don't want to make it harder to code review this PR.

@cachedout How about "freezing" this PR in terms of functionality; if any functional changes are pending or more come up, let's defer those to follow up PRs?

File: logstash-core/lib/logstash/config/lir_serializer.rb (hunk 17 above)
Comment (cachedout) [reply logstash-core/lib/logstash/config/lir_serializer.rb]: Hi @yaauie. Apologies for the confusion. This was a discussion [happening earlier in the PR discussion](https://github.com/elastic/logstash/pull/10576#issuecomment-498516575).

To avoid modifying the internal representation, do you think it would be better to try to do this lookup and insertion from [the API layer](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/api/commands/node.rb)?

**Comment by cachedout** [comment]
> @cachedout How about "freezing" this PR in terms of functionality; if any functional changes are pending or more come up, let's defer those to follow up PRs?

Agreed. Apologies, @yaauie for the moving target. We'll stick to just applying your review suggestions on this PR and will avoid any unexpected changes.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 18]
```diff
@@ -45,14 +60,25 @@ def events
           )
         end
 
-        def pipeline(pipeline_id = nil)
+        def pipeline(pipeline_id = nil, opts={})
+          extended_stats = LogStash::Config::PipelinesInfo.format_pipelines_info(
```
[the remark this answers is not in the document]
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: My apologies. I had conflated this change with another requested change. I have now implemented this as-written.

**Comment by cachedout** [comment]
> Some change in this PR seems to have broken the functionality of the graph query string parameter on the GET _node/pipelines API, introduced in #10561.

@ycombinator I have resolved this.

@yaauie FYI, so that you're aware, [the fix is in this commit](https://github.com/elastic/logstash/pull/10576/commits/bb748f80e4bcaebf906b5dfc56313db556464430).

**Comment by ycombinator** [comment]
> > Some change in this PR seems to have broken the functionality of the graph query string parameter on the GET _node/pipelines API, introduced in #10561.

> @ycombinator I have resolved this.

Confirmed that this is fixed again.

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 19]
```diff
@@ -121,8 +146,17 @@ def report(stats)
...
               :reloads => stats[:reloads],
               :queue => stats[:queue]
-            }.merge(stats[:dlq] ? {:dead_letter_queue => stats[:dlq]} : {})
+            }
+            ret[:dead_letter_queue] = stats[:dlq] if stats.include?(:dlq)
+
+            # if extended_stats were provided, enrich the return value
+            if extended_stats
+              ret[:queue]    = extended_stats["queue"] if extended_stats.include?("queue")
+              ret[:vertices] = extended_stats["vertices"] if opts[:vertices] && extended_stats.include?("vertices")
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]: I think this might be the ideal spot to inject the decoration of vertices with plugin metadata.

```suggestion
              if opts[:vertices] && extended_stats.include?("vertices")
                ret[:vertices] = extended_stats["vertices"].map { |vertex| decorate_vertex(vertex) }
              end
```

Which would work with the following private method. Note that we explicitly avoid mutating the vertices entries, as these are shared memory structures that outlive our individual API requests.

~~~ ruby
          private
          ##
          # Returns a vertex, decorated with additional metadata if available.
          # Does not mutate the passed `vertex` object.
          #
          # @api private
          # @param vertex [Hash{String=>Object}]
          # @return [Hash{String=>Object}]
          def decorate_vertex(vertex)
            plugin_id = vertex["id"]&.to_s
            return vertex unless plugin_id && LogStash::PluginMetadata.exists?(plugin_id)

            plugin_metadata = LogStash::PluginMetadata.for_plugin(plugin_id)
            cluster_uuid = plugin_metadata&.get(:cluster_uuid)
            vertex = vertex.merge("cluster_uuid" => cluster_uuid) unless cluster_uuid.nil?

            vertex
          end
~~~

File: logstash-core/lib/logstash/config/lir_serializer.rb (hunk 17 above)
Comment (yaauie) [reply logstash-core/lib/logstash/config/lir_serializer.rb]: Yes, I believe that cluster-uuid is an API-level concern and should be injected at that layer if feasible (whereas the other things that are injected here in `decorate_vertex` are all guaranteed to be available as soon as the pipeline compiles, the `cluster_uuid` requires further initialisation and connection to external resources).

I've added a suggestion as to how this could be added in `PluginStats#report`.

**Comment by cachedout** [comment]
Thanks for looking at this again, @yaauie. I have applied your suggestion.

File: logstash-core/lib/logstash/api/commands/stats.rb (hunk 19 above)
Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Hi @yaauie 

This does not seem to work for me when the `decorate_vertex` method is marked as `private`. 

Here is the changed section of the file:

```
        module PluginsStats
          module_function

          def plugin_stats(stats, plugin_type)
            # Turn the `plugins` stats hash into an array of [ {}, {}, ... ]
            # This is to produce an array of data points, one point for each
            # plugin instance.
            return [] unless stats[:plugins] && stats[:plugins].include?(plugin_type)
            stats[:plugins][plugin_type].collect do |id, data|
              { :id => id }.merge(data)
            end
          end

          private
          module_function
          ##
          # Returns a vertex, decorated with additional metadata if available.
          # Does not mutate the passed `vertex` object.
          #
          # @api private
          # @param vertex [Hash{String=>Object}]
          # @return [Hash{String=>Object}]
          def decorate_vertex(vertex)
            plugin_id = vertex["id"]&.to_s
            return vertex unless plugin_id && LogStash::PluginMetadata.exists?(plugin_id)

            plugin_metadata = LogStash::PluginMetadata.for_plugin(plugin_id)
            cluster_uuid = plugin_metadata&.get(:cluster_uuid)
            vertex = vertex.merge("cluster_uuid" => cluster_uuid) unless cluster_uuid.nil?

            vertex
          end

          def report(stats, extended_stats=nil, opts={})
            ret = {
              :events => stats[:events],
              :plugins => {
                :inputs => plugin_stats(stats, :inputs),
                :codecs => plugin_stats(stats, :codecs),
                :filters => plugin_stats(stats, :filters),
                :outputs => plugin_stats(stats, :outputs)
              },
              :reloads => stats[:reloads],
              :queue => stats[:queue]
            }
            ret[:dead_letter_queue] = stats[:dlq] if stats.include?(:dlq)

            # if extended_stats were provided, enrich the return value
            if extended_stats
              ret[:queue]    = extended_stats["queue"] if extended_stats.include?("queue")
              if opts[:vertices] && extended_stats.include?("vertices")
                ret[:vertices] = extended_stats["vertices"].map { |vertex| decorate_vertex(vertex) }
              end
            end

<snip>
```

It does work correctly when marking `decorate_vertex` as a `private_class_method` instead of simply `private`, but I'm afraid I don't know if this achieves the effect you were intending.

The good news is that this does work correctly when the `private` marker is removed.

Comment (yaauie) [reply logstash-core/lib/logstash/api/commands/stats.rb]: ah. I missed the `Module#module_function` declaration at the top of the module (which makes the subsequently-defined instance methods private and also extends them publicly onto the module itself), so we can just avoid sending `Module#private` entirely.

***

Since `PluginStats#decorate_vertex` is an _implementation detail_ defined to make `PluginStats#report` more readable, it is effectively a private-API method and should be declared at the _bottom_ of this module definition.

While you may see plenty of evidence to the contrary in this project, by convention in Ruby, we tend to declare methods in an order that indicates publicity instead of oscillating between modes: first all of the public methods, then protected, then private. This allows a reader to quickly understand a module or class's public API quickly, to be aware of the scope of effect of a change to a class's implementation, and to avoid tying one class's behaviour to another class's implementation details.

Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Hi @yaauie. I've removed the `private` declaration and I believe that's the last outstanding issue on this PR. Thanks!

File: logstash-core/lib/logstash/api/commands/stats.rb [hunk 20]
```diff
@@ -110,8 +135,25 @@ def plugin_stats(stats, plugin_type)
             end
           end
 
-          def report(stats)
-            {
+          ##
+          # Returns a vertex, decorated with additional metadata if available.
+          # Does not mutate the passed `vertex` object.
+          #
```
Comment (yaauie) [inline logstash-core/lib/logstash/api/commands/stats.rb]:
```suggestion
          #
          # @api private
```

Even though this method must be public to avoid restructuring of the module, the `@api private` comment should remain in order to discourage developers from using it from outside of the module.

Additionally, per my previous comment and by convention, this private-API method definition should go at the _bottom_ of the module definition, below the public-API methods.

> While you may see plenty of evidence to the contrary in this project, by convention in Ruby, we tend to declare methods in an order that indicates publicity instead of oscillating between modes: first all of the public methods, then protected, then private. This allows a reader to quickly understand a module or class's public API quickly, to be aware of the scope of effect of a change to a class's implementation, and to avoid tying one class's behaviour to another class's implementation details.
>
> -- https://github.com/elastic/logstash/pull/10576#discussion_r295419677

Comment (cachedout) [reply logstash-core/lib/logstash/api/commands/stats.rb]: Thanks, @yaauie. I have replaced the @private marker in the method documentation and moved it to the bottom.

**Comment by cachedout** [comment]
Thanks again for your help @yaauie. It's very much appreciated by me and the whole team. :)

@ycombinator Just as an FYI, the backport to this is here: https://github.com/elastic/logstash/pull/10908
